src/frequenz/sdk/actor/data_sourcing/microgrid_source.py

```diff
@@ -0,0 +1,373 @@
+"""The Microgrid API data source for the DataSourcingActor.
+
+Copyright
+Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+License
+MIT
+"""
+
+import asyncio
+from dataclasses import dataclass
+from datetime import datetime
+from typing import Any, Callable, Dict, List, Optional, Tuple
+
+from frequenz.channels import Receiver, Sender
+
+from frequenz.sdk.actor import ChannelRegistry
+from frequenz.sdk.microgrid import (
+    BatteryData,
+    ComponentCategory,
+    EVChargerData,
+    InverterData,
+    MeterData,
+    microgrid_api,
+)
+
+from .common import ComponentMetric, ComponentMetricRequest
+
+_MeterDataMethods: Dict[ComponentMetric, Callable[[MeterData], float]] = {
+    ComponentMetric.ACTIVE_POWER: lambda msg: msg.active_power,
+    ComponentMetric.CURRENT_PHASE_1: lambda msg: msg.current_per_phase[0],
+    ComponentMetric.CURRENT_PHASE_2: lambda msg: msg.current_per_phase[1],
+    ComponentMetric.CURRENT_PHASE_3: lambda msg: msg.current_per_phase[2],
+    ComponentMetric.VOLTAGE_PHASE_1: lambda msg: msg.voltage_per_phase[0],
+    ComponentMetric.VOLTAGE_PHASE_2: lambda msg: msg.voltage_per_phase[1],
+    ComponentMetric.VOLTAGE_PHASE_3: lambda msg: msg.voltage_per_phase[2],
+}
+
+_BatteryDataMethods: Dict[ComponentMetric, Callable[[BatteryData], float]] = {
+    ComponentMetric.SOC: lambda msg: msg.soc,
+    ComponentMetric.SOC_LOWER_BOUND: lambda msg: msg.soc_lower_bound,
+    ComponentMetric.SOC_UPPER_BOUND: lambda msg: msg.soc_upper_bound,
+    ComponentMetric.CAPACITY: lambda msg: msg.capacity,
+    ComponentMetric.POWER_LOWER_BOUND: lambda msg: msg.power_lower_bound,
+    ComponentMetric.POWER_UPPER_BOUND: lambda msg: msg.power_upper_bound,
+}
+
+_InverterDataMethods: Dict[ComponentMetric, Callable[[InverterData], float]] = {
+    ComponentMetric.ACTIVE_POWER: lambda msg: msg.active_power,
+    ComponentMetric.ACTIVE_POWER_LOWER_BOUND: lambda msg: msg.active_power_lower_bound,
+    ComponentMetric.ACTIVE_POWER_UPPER_BOUND: lambda msg: msg.active_power_upper_bound,
+}
+
+_EVChargerDataMethods: Dict[ComponentMetric, Callable[[EVChargerData], float]] = {
+    ComponentMetric.ACTIVE_POWER: lambda msg: msg.active_power,
+    ComponentMetric.CURRENT_PHASE_1: lambda msg: msg.current_per_phase[0],
+    ComponentMetric.CURRENT_PHASE_2: lambda msg: msg.current_per_phase[1],
+    ComponentMetric.CURRENT_PHASE_3: lambda msg: msg.current_per_phase[2],
+    ComponentMetric.VOLTAGE_PHASE_1: lambda msg: msg.voltage_per_phase[0],
+    ComponentMetric.VOLTAGE_PHASE_2: lambda msg: msg.voltage_per_phase[1],
+    ComponentMetric.VOLTAGE_PHASE_3: lambda msg: msg.voltage_per_phase[2],
+}
+
+
+## FIXME: does `src/frequenz/sdk/data_pipeline/types.py` sounds like a good
+## place for the `Sample` class?
+@dataclass(frozen=True)
+class Sample:
+    """Describes a single observed floating point value at a specific timestamp."""
```

Comment:
As I said in the resampling PR, maybe we can just say `measurement` to be more consistent with the glossary. Also maybe mention why and when the value could be `None`:

```suggestion
    """A measurement taken at a particular point in time.
    
    The `value` could be `None` if a component is malfunctioning or data is lacking
    for another reason, but a sample still needs to be sent to have a coherent view
    on a group of component metrics for a particular timestamp.
    """
```

Reply:
done